src/main/java/com/metamx/metrics/JvmMonitor.java

```diff
@@ -206,14 +115,165 @@ public boolean doMonitor(ServiceEmitter emitter)
+        final ServiceMetricEvent.Builder builder = new ServiceMetricEvent.Builder();
+        MonitorUtils.addDimensionsToBuilder(builder, dimensions);
+        generation.emit(emitter, builder);
+      }
+    }
+  }
+
+  private class GcGeneration
   {
-    final ServiceMetricEvent.Builder builder = new ServiceMetricEvent.Builder();
-    MonitorUtils.addDimensionsToBuilder(builder, dimensions);
-    gcCounters.emitCounters(emitter, builder);
+    private String name;
+    private GcGenerationCollector collector = null;
+    private List<GcGenerationSpace> spaces = new ArrayList<>();
+
+    GcGeneration(Map<String, JStatData.Counter<?>> jStatCounters, long genIndex, String name)
+    {
+      this.name = name.toLowerCase();
+
+      long spacesCount = ((JStatData.LongCounter) jStatCounters.get(String.format("sun.gc.generation.%d.spaces", genIndex))).getLong();
+      for (long spaceIndex = 0; spaceIndex < spacesCount; spaceIndex++) {
+        spaces.add(new GcGenerationSpace(jStatCounters, genIndex, spaceIndex));
+      }
+
+      if (jStatCounters.containsKey(String.format("sun.gc.collector.%d.name", genIndex))) {
+        collector = new GcGenerationCollector(jStatCounters, genIndex);
+      }
+    }
+
+    void emit(ServiceEmitter emitter, ServiceMetricEvent.Builder builder)
+    {
+      builder.setDimension("gcGen", name);
+
+      if (collector != null) {
+        collector.emit(emitter, builder);
+      }
+
+      for (GcGenerationSpace space: spaces) {
+        space.emit(emitter, builder);
+      }
+    }
+  }
+
+  private class GcGenerationCollector
+  {
+    private String name;
+    private final LongCounter invocationsCounter;
+    private final TickCounter cpuCounter;
+    private long lastInvocations = 0;
+    private long lastCpuNanos = 0;
+
+    GcGenerationCollector(Map<String, JStatData.Counter<?>> jStatCounters, long genIndex)
+    {
+      String collectorKeyPrefix = String.format("sun.gc.collector.%d", genIndex);
+
+      String nameKey = String.format("%s.name", collectorKeyPrefix);
+      StringCounter nameCounter = (StringCounter) jStatCounters.get(nameKey);
+      name = getReadableName(nameCounter.getString());
+
+      invocationsCounter = (LongCounter) jStatCounters.get(String.format("%s.invocations", collectorKeyPrefix));
+      cpuCounter = (TickCounter) jStatCounters.get(String.format("%s.time", collectorKeyPrefix));
+    }
+
+    void emit(ServiceEmitter emitter, ServiceMetricEvent.Builder builder)
+    {
+      builder.setDimension("gcName", name);
+
+      long newInvocations = invocationsCounter.getLong();
+      emitter.emit(builder.build("jvm/gc/count", newInvocations - lastInvocations));
+      lastInvocations = newInvocations;
+
+      long newCpuNanos = cpuCounter.getNanos();
+      emitter.emit(builder.build("jvm/gc/cpu", newCpuNanos - lastCpuNanos));
+      lastCpuNanos = newCpuNanos;
+    }
+
+    private String getReadableName(String name)
+    {
+      switch (name) {
+        // Young gen
```

Comment:
I kept this switch like this to show a difference between young and old generations. If you think it's better to group them not by generation but by GC I can refactor that